primary/src/block_synchronizer/mod.rs

```diff
@@ -0,0 +1,756 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    block_synchronizer::{
+        peers::Peers,
+        responses::{CertificatesResponse, PayloadAvailabilityResponse, RequestID},
+        PendingIdentifier::{Header, Payload},
+    },
+    primary::PrimaryMessage,
+    utils, BatchDigest, Certificate, CertificateDigest, PayloadToken, PrimaryWorkerMessage,
+};
+use bytes::Bytes;
+use config::{BlockSynchronizerParameters, Committee, WorkerId};
+use crypto::{traits::VerifyingKey, Hash};
+use futures::{
+    future::{join_all, BoxFuture},
+    stream::FuturesUnordered,
+    FutureExt, StreamExt,
+};
+use network::SimpleSender;
+use rand::{rngs::SmallRng, SeedableRng};
+use std::{collections::HashMap, time::Duration};
+use store::Store;
+use thiserror::Error;
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    time::{sleep, timeout},
+};
+use tracing::log::{debug, error, warn};
+
+#[cfg(test)]
+#[path = "tests/block_synchronizer_tests.rs"]
+mod block_synchronizer_tests;
+mod peers;
+mod responses;
+
+/// The minimum percentage
+/// (number of responses received from primary nodes / number of requests sent to primary nodes)
+/// that should be reached when requesting the certificates from peers in order to
+/// proceed to next state.
+const CERTIFICATE_RESPONSES_RATIO_THRESHOLD: f32 = 0.5;
+
+type ResultSender<T> = Sender<BlockSynchronizeResult<Certificate<T>>>;
+type BlockSynchronizeResult<T> = Result<T, SyncError>;
+
+pub enum Command<PublicKey: VerifyingKey> {
```

Comment:
Very good point. When building this component my assumption was that this would be the client's/consumer's responsibility. This component wouldn't look into the data storage directly. Anyone could use the `block_synchronizer` to sync or (re-sync for whatever reason) the blocks. Now, that being said I thinking whether mainly _for convenience reasons_ we should embed those checks into this component . For the moment. I'll add a comment to make this clear and think on iterating on it.